File: code/validate.py
#!/usr/bin/env python3
import sys
import os.path
import json
import jsonschema
from referencing import Registry, Resource


schema_uri_prefix = 'https://burrito.bible/schema/'
schema_dir = os.path.join(os.path.dirname(os.path.abspath(__file__)), '..', 'schema')
schema = json.load(open(os.path.join(schema_dir, 'metadata.schema.json')))


def http_handler(uri):
    """Resolves URLs like https://burrito.bible/schema/... to disk"""

    assert uri.startswith(schema_uri_prefix)
    path = uri[len(schema_uri_prefix):].replace('/', os.sep)
    path = os.path.join(schema_dir, path)
    return Resource.from_contents(json.load(open(path, encoding='utf-8')))


registry = Registry(retrieve=http_handler)


def validate(input):
    "Validates the given file-like object against the schema."
    data = json.load(input)
    jsonschema.validate(data, schema, registry=registry)


if __name__ == '__main__':
    files = sys.argv[1:]
    inputs = []
    for file in files:
        if file == '-':
            inputs.append(sys.stdin)
        else:
            inputs.append(open(file, 'r'))

    if not inputs:
        inputs.append(sys.stdin)

    success = True

    for input in inputs:
        try:
            validate(input)
            print("{0}: No errors.".format(input.name))
        except Exception as ex:
            print("{0}: {1}".format(input.name, str(ex)))
            success = False

    if not success:
        sys.exit(1)

File: code/validate.js
const schemaIndex = require("../schema");
const fs = require("fs");
const globSync = require("glob").globSync;
const path = require("path");
const Ajv = require("ajv");
const addFormats = require("ajv-formats");

const ajv = new Ajv({ schemas: schemaIndex.schemas });

ajv.addKeyword("$$target");
ajv.addKeyword("revision");
ajv.addKeyword("timestamp");
addFormats(ajv);

function validate(schemaName, fn, data) {
  const validator = ajv.getSchema(schemaIndex.schemaIds[schemaName]);
  if (validator(JSON.parse(data))) {
    console.log(fn + ": No errors.");
    return true;
  } else {
    console.log(fn + ":", validator.errors);
    return false;
  }
}

function FindFiles(pattern) {
  const fullPath = path.join(__dirname, "..", pattern);
  const fileDir = path.dirname(fullPath);
  const files = globSync(path.basename(fullPath), { nodir: true, cwd: fileDir });
  return files.map(file => path.join(fileDir, file));
}

let success = true;

if (process.argv.length > 3) {
  for (const arg of process.argv.slice(3)) {
    for (const fPat of FindFiles(arg)) {
      const fNm = path.basename(fPat);
      if (!validate(process.argv[2], fNm, fs.readFileSync(fPat, "utf-8"))) {
        success = false;
      }
    }
  }
} else {
  const data = fs.readFileSync(0, "utf-8");
  success = validate(process.argv[2], "<stdin>", data);
}
if (!success) {
  process.exit(1);
}
